### PR TITLE
Implemented Phone Number Verification and Bug Fixes

### PR DESCRIPTION
**Pull request by rs929**
## Overview

- Implemented Phone Number Verification for Scooped Onboarding
- Onboarding Bug Fixes

## Changes Made

- Implemented Phone Number Verification with Firebase
- Created new VerifyPhoneNumberVC
- On verification, a 6 -digit verification code is sent to the user's phone number and they must input the code to continue with onboarding
- Fixed bug where whitespace before and after textfields were not stripped
- Fixed bug where sliders would always revert back to first value on click

## Related PRs or Issues

https://github.com/cuappdev/scoop-ios/issues/42
https://github.com/cuappdev/scoop-ios/issues/43
https://github.com/cuappdev/scoop-ios/issues/45

## Screenshots

<video src = https://user-images.githubusercontent.com/105038960/234467192-cad64059-ec52-46bd-baa6-b523e433fc05.mp4 width = "300px" height = "auto">

### CODE REVIEW

File: Scoop/Controller/Onboarding/FavoritesViewController.swift [hunk 1]
```diff
@@ -33,10 +33,12 @@ class FavoritesViewController: OnboardingViewController {
                       return
                   }
             
-            self.addPrompt(name: "Snack", placeholder: "Chips", answer: snack)
-            self.addPrompt(name: "Song", placeholder: "Favorite Song", answer: song)
-            self.addPrompt(name: "Stop", placeholder: "Gates Hall", answer: stop)
-            
+            let snackTrimmed = snack.trimmingCharacters(in: .whitespacesAndNewlines)
```
Comment (tiffany-pann) [inline Scoop/Controller/Onboarding/FavoritesViewController.swift:36]: maybe use a forEach here

File: Scoop/Controller/Onboarding/OnboardingViewController.swift [hunk 2]
```diff
@@ -140,7 +140,7 @@ class OnboardingViewController: UIViewController {
                 
                 print("Updated: \(prompt.questionName) to \(answer)")
             case .failure(let error):
-                print(error.localizedDescription)
+                print("Error in OnboardingViewController: \(error.localizedDescription)")
```
Comment (tiffany-pann) [inline Scoop/Controller/Onboarding/OnboardingViewController.swift:159]: nice!

File: Scoop/Controller/LoginViewController.swift [hunk 3]
```diff
@@ -106,25 +106,25 @@ class LoginViewController: UIViewController {
     }
     
     private func getUser() {
-        NetworkManager.shared.getUser { result in
+        NetworkManager.shared.getUser { [weak self] result in
```
Comment (ReadeP) [inline Scoop/Controller/LoginViewController.swift:109]: beautiful <3

File: Scoop/Controller/LoginViewController.swift [hunk 4]
```diff
@@ -106,25 +106,25 @@ class LoginViewController: UIViewController {
...
-                    return
-                }
-                
+                guard let window = UIApplication.shared.windows.first else { return }
                 NotificationCenter.default.post(name: NSNotification.Name("didCompleteLogin"), object: nil)
+                
                 if GIDSignIn.sharedInstance.hasPreviousSignIn() {
                     GIDSignIn.sharedInstance.signOut()
-                    self.presentErrorAlert(title: "Failed to Login", message: "Please make sure you have a stable internet connection and try again later")
+                    strongSelf.presentErrorAlert(title: "Failed to Login", message: "Please make sure you have a stable internet connection and try again later")
```
Comment (ReadeP) [inline Scoop/Controller/LoginViewController.swift]: Add a period at the end of "Please make sure you have a stable internet connection and try again later"

File: Scoop/Controller/Onboarding/PreferredContactViewController.swift [hunk 5]
```diff
@@ -195,3 +203,18 @@ extension PreferredContactViewController: UITextFieldDelegate {
...
+    
+    func verificationSuccess(phoneNumber: String) {
+        isVerified = true
+        NetworkManager.shared.currentUser.phoneNumber = phoneNumber
+        guard let navCtrl = self.navigationController else {
+            return
+        }
+        
+        delegate?.didTapNext(navCtrl, nextViewController: nil)
+        print("UPDATED")
```
Comment (ReadeP) [inline Scoop/Controller/Onboarding/PreferredContactViewController.swift]: delete this print statement

File: Scoop/Controller/Onboarding/VerifyPhoneNumberViewController.swift [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+        self.phoneNumber = phoneNumber
+        self.phoneDelegate = delegate
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(true)
```
Comment (ReadeP) [inline Scoop/Controller/Onboarding/VerifyPhoneNumberViewController.swift]: `super.viewDidAppar(animated)`

File: Scoop/Controller/Onboarding/VerifyPhoneNumberViewController.swift [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+    }
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        view.backgroundColor = .white
+        self.navigationItem.hidesBackButton = true
```
Comment (ReadeP) [inline Scoop/Controller/Onboarding/VerifyPhoneNumberViewController.swift]: remove `self`

File: Scoop/Controller/Onboarding/VerifyPhoneNumberViewController.swift [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+                guard let phoneNumber = self.phoneNumber else { return }
+                
+                self.navigationController?.popViewController(animated: true)
+                self.phoneDelegate?.verificationSuccess(phoneNumber: phoneNumber)
+            }
+            
+            alertVC.addAction(okAction)
+            alertVC.view.tintColor = .scoopDarkGreen
+            self.present(alertVC, animated: true)
+            print("SUCCESS")
```
Comment (ReadeP) [inline Scoop/Controller/Onboarding/VerifyPhoneNumberViewController.swift]: delete print statement

**Review by ReadeP — APPROVED** [review]
Amazing! :D

File: Scoop/Controller/Onboarding/VerifyPhoneNumberViewController.swift [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+            make.width.equalTo(120)
+            make.height.equalTo(56)
+            make.center.equalToSuperview()
+        }
+    }
+    
+}
+
+extension VerifyPhoneNumberViewController: UITextFieldDelegate {
+    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
```
Comment (tiffany-pann) [inline Scoop/Controller/Onboarding/VerifyPhoneNumberViewController.swift:187]: i think usually there is a linebreak in between the first function?

File: Scoop/Controller/Onboarding/VerifyPhoneNumberViewController.swift [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+        let numOnly = CharacterSet.decimalDigits
+        let characters = CharacterSet(charactersIn: string)
+        // Source: https://stackoverflow.com/a/31363255/5278889
+        let maxLength = 6
+        let currentString: NSString = (textField.text ?? "") as NSString
+        let newString: NSString =
+        currentString.replacingCharacters(in: range, with: string) as NSString
+        return newString.length <= maxLength && numOnly.isSuperset(of: characters)
+    }
```
Comment (tiffany-pann) [inline Scoop/Controller/Onboarding/VerifyPhoneNumberViewController.swift:196]: and then a line break here as well

**Review by tiffany-pann — APPROVED** [review]
sick work richie! you are insane 🚀

**Pull request merged**